builder/codebuild/codebuild.go

```diff
@@ -0,0 +1,353 @@
+		ProjectName: ProjectName,
+		DockerCfg:   "conveyor.dockercfg",
+		codebuild:   codebuild.New(config),
+		logReader:   logReader,
+	}
+}
+
+// Build executes the docker image.
+func (b *Builder) Build(ctx context.Context, w io.Writer, opts builder.BuildOptions) (image string, err error) {
+	projectName := b.ProjectName(opts)
+	image = fmt.Sprintf("%s:%s", opts.Repository, opts.Sha)
+	err = b.build(ctx, w, BuildOptions{
+		BuildOptions: opts,
+		ProjectName:  projectName,
+		DryRun:       b.DryRun,
+		DockerCfg:    b.DockerCfg,
+	})
+	return
+}
+
+func (b *Builder) build(ctx context.Context, w io.Writer, opts BuildOptions) error {
+	resp, err := b.startBuild(ctx, opts)
+	if err != nil {
+		// if the project does not exist, create it.
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "ResourceNotFoundException" {
+			if _, err := b.createProject(ctx, opts); err != nil {
+				return err
+			}
+
+			resp, err = b.startBuild(ctx, opts)
+			if err != nil {
+				return err
+			}
+		} else {
+			return fmt.Errorf("starting build for %s: %v", opts.Repository, err)
+		}
+	}
+
+	buildID := resp.Build.Id
+
+	done := make(chan error)
+	go func() {
+		done <- b.tailBuild(ctx, w, buildID)
+	}()
+
+	select {
+	case <-ctx.Done():
+		// Try cancelling the build. We don't really care about the
+		// error here.
+		b.codebuild.StopBuild(&codebuild.StopBuildInput{
+			Id: buildID,
+		})
+
+		// No wait for log streaming and everything to finish up.
+		err = <-done
+		if err == nil {
+			err = ctx.Err()
+		}
+	case err = <-done:
+	}
+
+	if err == context.Canceled || err == context.DeadlineExceeded {
+		return &builder.BuildCanceledError{
+			Err:    err,
+			Reason: ctx.Err(),
+		}
+	}
+
+	return err
+}
+
+func (b *Builder) tailBuild(ctx context.Context, w io.Writer, buildID *string) error {
+	fmt.Fprintf(w, "conveyor: Waiting for CodeBuild build %s to start\n", *buildID)
+
+	build, err := b.waitUntilLogsAvailable(ctx, buildID)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintf(w, "conveyor: CodeBuild build %s started, streaming logs...\n", *buildID)
+
+	r, err := b.logReader(*build.Logs.GroupName, *build.Logs.StreamName)
+	if err != nil {
+		return fmt.Errorf("opening cloudwatch logs stream for %s: %v", *build.Id, err)
+	}
+
+	copyDone := make(chan error)
+	go func() {
+		_, err := io.Copy(w, r)
+		copyDone <- err
+	}()
+
+	ctxTimeout, cancel := context.WithTimeout(context.Background(), BuildCompleteTimeout)
+	defer cancel()
+
+	build, _ = b.waitUntilBuildComplete(ctxTimeout, build.Id)
+
+	// Close the cloudwatch log stream reader.
+	r.Close()
+
+	// Wait for the io.Copy to complete flushing data to w.
+	if err := <-copyDone; err != nil {
+		return fmt.Errorf("copying logs: %v", err)
+	}
+
+	fmt.Fprintf(w, "conveyor: CodeBuild build %s completed\n", *buildID)
+
+	for _, status := range failedStatuses {
+		if *build.BuildStatus == status {
+			return fmt.Errorf("CodeBuild build %s failed", *buildID)
+		}
+	}
```

Comment:
If the status is not in the failed list but unknown, it's not acted upon (at least locally) -- is it recovered from somewhere else? In other case, it could be good to log it anyway in case new statuses appear.